docs/autorenew-feature.md

```diff
@@ -0,0 +1,82 @@
+
+## Design
+- Introduce new settings in `application.properties`:
+  * `autorenew.isEnabled`
+  * `autorenew.numberOfEntitiesToScan`
+  * `autorenew.maxNumberOfEntitiesToRenewOrDelete`
+  * `autorenew.gracePeriod`
+- Each Hedera entity has an `expirationTime` which is the effective consensus timestamp at (and after) which the entity is set to expire.
+- Each Hedera entity also has an `autoRenewAccount` which is the account to pay for the fees at renewal. A crypto account or a smart contract is its own `autoRenewAccount` account.
+- When a Hedera entity is created, the payer account is charged enough hbars (as a rental fee) for the entity to stay active in the ledger state until consensus time passes its `expirationTime`. At its expiration time, Hedera Services will try to extend an entity's expiration time by another `autoRenewPeriod` if the `autoRenewAccount` has enough balance to do so, or as much extension as the remaining balance permits.
+- `autoRenewPeriod` is a field to customize the extension period in seconds for a crypto account, a topic, a smart contract, or a token type. For a file, the extension period will be three months. Future protobuf changes will permit customizing this extension period as well. Entities associated with new services in the future will also have an `autoRenewPeriod` and an associated autorenew account.
+- A Hedera entity that lacks a funded `autoRenewAccount`, namely `autoRenewAccount` is not specified or has zero balance at the time renewal fees are due, will be marked as expired. The entity will then have a grace period (defaulted to 7 days) before removal (permanent deletion). At the end of the grace period, it will again try to autorenew, but if there is no account, or the balance is zero, then it will be permanently removed at that time.
+- A Hedera entity can have its expiration time extended by anyone, not just by the owner or admin account. The expiration time is the only field that can be changed in an update without being signed by the owner or the admin. It is also the only field that can be changed while expired (during the grace period).
+- Every token type has an associated account as its treasury account. An expired account that is permanently removed (not being renewed after the grace period) will have its tokens transfer to the treasury account. The treasury account always has an expiration time the same as or later than its token type, because extending the latter will automatically extend the former, if needed. So the treasury account can’t expire unless the token type itself expires, in which case, all tokens of that type expire.
+- After handling a transaction, all Hedera Services nodes will perform a synchronous scanning of the next `autorenew.numberOfEntitiesToScan` for up to `autorenew.maxNumberOfEntitiesToRenewOrDelete` entities that expired, then try to autorenew these entities.
```

Comment:
After every transaction seem performance heavy.
Might it be better to do this at a predictable rate. Either every x transactions or every x time period and in a process that doesn't affect event/transaction processing.
This might be an implementation detail.